### ISSUE TITLE
ML training scripts: portable checkpoint-resume pattern across all sweeps

### ISSUE THREAD

**Issue by jsboige** [jsboige/CoursIA#1053]
## Context

Le restart VSCode du 2026-05-14 ~00:48Z a tue M15 LSTM h=128 BG (PID 23664) qui tournait depuis ~3h sans aucun checkpoint persiste. Toute la progression (9/84 combos) a ete perdue car le script ne sauvait `results.json` qu'apres la fin complete des 84 combos.

Pattern de checkpoint-resume ajoute a `m15_lstm_rv.py` (commit local, branche `feat/m15-lstm-rv`) :
- Append-per-combo dans `results/<sweep>/checkpoint.jsonl` apres chaque `evaluate_one_combo` reussi
- Au demarrage : lit le checkpoint, build `completed_keys: set[tuple]` sur `(coin, horizon, seed)`, skip les combos deja faits
- Granularite : 1 combo = ~3-4 min, donc restart = perte max 1 combo

## Scope (etendre le pattern)

Tous les scripts de sweep qui itèrent sur une grille (`coin x horizon x seed` ou variantes) sans persistance intermediaire actuellement. Liste a auditer :

### Sweeps actifs

- [ ] `m11d_hmm_kelly.py` (HMM PyMC, walk-forward)
- [ ] `m11g_fee_aware_kelly.py`
- [ ] `m11h_kelly_cap_relaxed.py`
- [ ] `m11j_multi_asset_kelly.py`
- [ ] `m12_har_rv_j.py` (HAR-RV-J per-coin)
- [ ] `m13_ms_har.py` (MS-HAR Markov-switching)
- [ ] `m14_heavy.py` (HEAVY model)
- [ ] `m15_lstm_rv.py` **DONE** (reference implementation)
- [ ] M16/M17/M18 (a creer)

### Architectures SOTA (M8 family, sweep 96 runs deja avorte sans checkpoint)

- [ ] `train_tft.py`
- [ ] `train_mamba.py`
- [ ] `train_itransformer.py`
- [ ] `train_patchtst.py`
- [ ] `train_transformer.py`
- [ ] `train_lstm.py` (legacy, peut-etre obsolete maintenant)

### Specialises

- [ ] `train_moe.py`, `train_moe_regimes.py` (MoE, sweep regimes)
- [ ] `train_volatility_garch_dl.py`, `train_realized_garch.py`
- [ ] `train_har_baseline.py`
- [ ] `train_dqn_rl.py`, `train_rl_dt.py`
- [ ] `train_classification.py`, `train_heteroscedastic.py`, `train_multiscale_gnn.py`, `train_gnn.py`, `train_stgat.py`, `train_mtgnn.py`
- [ ] `train_recursive_multistep.py`
- [ ] `train_ensemble.py`
- [ ] `train_regime_classifier.py`, `train_volatility_regime.py`
- [ ] `train_baselines_crypto_panier.py`

## Pattern de reference (m15_lstm_rv.py)

```python
results_dir = SCRIPT_DIR / "results" / f"<sweep_name>"
results_dir.mkdir(parents=True, exist_ok=True)
checkpoint_path = results_dir / "checkpoint.jsonl"

combos: list[dict] = []
completed_keys: set[tuple] = set()
if checkpoint_path.exists():
    with open(checkpoint_path, "r") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            row = json.loads(line)
            combos.append(row)
            completed_keys.add((row["coin"], row["horizon"], row["seed"]))  # ADAPT KEY
    print(f"[CHECKPOINT] resumed {len(combos)} combos from {checkpoint_path.name}", flush=True)

for coin in COINS:
    for h in HORIZONS:
        for seed in SEEDS:
            key = (coin, h, seed)
            if key in completed_keys:
                print(f"[{done}/{total}] {coin} h={h} seed={seed} -- SKIP (checkpoint)")
                continue
            row = evaluate_one_combo(coin, h, seed, ...)
            if row is not None:
                combos.append(row)
                with open(checkpoint_path, "a") as f:
                    f.write(json.dumps(row, default=str) + "\n")
```

### Variantes a gerer par sweep

- Cle differente : `(coin,)` seul pour M12 per-coin, `(coin, regime)` pour MS-HAR, `(coin, fee)` pour M11f tx-cost sweep, etc. Adapter `completed_keys` au tuple effectif.
- Sweep epoch-based (DQN, NNs) : ajouter aussi un checkpoint torch (`model_state_dict`) toutes les N epochs OU avant abort sigterm handler.
- Sweep avec pre-computation lourde (data loading, indicator caching) : sauver le cache avant la boucle pour eviter de tout recharger au resume.

## Acceptance criteria

- [ ] Tous les scripts de sweep listes ci-dessus ont le pattern checkpoint-resume
- [ ] Test E2E : lancer le sweep, kill apres N combos, relancer, verifier que les N premiers sont skip et que le sweep continue
- [ ] `checkpoint.jsonl` non commite (ajouter `**/checkpoint.jsonl` au `.gitignore` si pas deja)
- [ ] Documentation : section "Checkpoint-resume" dans `MyIA.AI.Notebooks/QuantConnect/ML-Training-Pipeline/scripts/README.md` (ou creer si absent)

## Bonus (separe, faible priorite)

- [ ] Helper module `checkpoint_utils.py` (load_checkpoint / append_to_checkpoint) pour eviter la duplication
- [ ] SIGTERM handler propre pour flush avant kill (utile pour le restart VSCode automatique)
- [ ] Reporting : log `[CHECKPOINT] N/Total skip + M/Total resume` au demarrage de chaque sweep

## Dispatch

- Cluster ai-01 : owner du ML training pipeline. Faisable progressif a chaque relance de sweep (`m12` puis `m13` puis `m14` puis SOTA family).
- Pas de deadline dure : le pattern m15 deja prouve resilient au prochain restart, les autres sont a faire au fil des relances.

Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

**Comment by jsboige** [comment]
## [DIRECTIVE ai-01 — auto-pickup] Checkpoint-resume burndown

Audit + harmonize the checkpoint-resume pattern that `m15_lstm_rv.py` already implements, across the rest of `MyIA.AI.Notebooks/QuantConnect/ML-Training-Pipeline/scripts/m*.py`.

### Scope

1. Inventory all M-series sweep scripts (m7_* through m17_*, m8_aggregate_*, moe_experts*, multiscale_*) — for each, check whether `results/<sweep>/checkpoint.jsonl` is appended per-combo and resumed on startup.
2. Identify scripts with NO checkpoint persistence (probable victims: m8_aggregate_*, multiscale_gnn_*, possibly older m11* helpers if they ever re-run).
3. Port the same minimal pattern from `m15_lstm_rv.py` lines 501–514 (append per combo, `(coin, horizon, seed)` key dedup on resume) to each gap script.

### Acceptance criteria

- `grep -l "checkpoint.jsonl" m*.py` returns all sweep scripts (not just M11+ / M12 / M13 / M14 / M15 / M17).
- Each script lists `[CHECKPOINT] resumed N combos` on startup when re-launched.
- Tested: kill mid-run, restart → resumes at previous combo, doesn't re-run completed ones.
- Single PR per ~5 scripts (rule G.4 split threshold), titled `feat(ml): checkpoint-resume for <list>`.

Do NOT touch m11ef / m13_ms_har (already shipped). Do NOT add new features beyond resume — no logging refactor, no progress bar, just the resume primitive.
